src/object/obj_tx.c

```diff
@@ -0,0 +1,1039 @@
+	dtsr->dtsr_dkey_hash = obj_dkey2hash(dkey);
+	dtsr->dtsr_flags = flags & ~DAOS_ZERO_COPY;
+
+	if (flags & DAOS_ZERO_COPY) {
+		dtsr->dtsr_zero_copy = 1;
+		dtsr->dtsr_dkey = dkey;
+	} else {
+		dtsr->dtsr_dkey = &dtsr->dtsr_dkey_inline;
+		rc = daos_iov_copy(dtsr->dtsr_dkey, dkey);
+		if (rc != 0)
+			D_GOTO(out, rc);
+	}
+
+	d_list_add_tail(&dtsr->dtsr_link, &tx->tx_sub_reqs);
+	tx->tx_sub_count++;
+
+out:
+	dc_tx_decref(tx);
+	if (rc != 0 && dtsr != NULL) {
+		if (dtsr->dtsr_obj != NULL)
+			obj_decref(dtsr->dtsr_obj);
+
+		D_FREE(dtsr);
+	}
+
+	return rc;
+}
+
+static int
+dc_tx_punch_akeys_attach(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+			 daos_key_t *dkey, unsigned int nr, daos_key_t *akeys)
+{
+	struct dc_tx		*tx = NULL;
+	struct dc_tx_sub_req	*dtsr = NULL;
+	int			 rc;
+	int			 i;
+
+	rc = dc_tx_check(th, true, &tx);
+	if (rc != 0)
+		return rc;
+
+	D_ALLOC_PTR(dtsr);
+	if (dtsr == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	dtsr->dtsr_obj = obj_hdl2ptr(oh);
+	if (dtsr->dtsr_obj == NULL)
+		D_GOTO(out, rc = -DER_NO_HDL);
+
+	dtsr->dtsr_opc = DAOS_OBJ_RPC_PUNCH_AKEYS;
+	dtsr->dtsr_dkey_hash = obj_dkey2hash(dkey);
+	dtsr->dtsr_flags = flags & ~DAOS_ZERO_COPY;
+	dtsr->dtsr_nr = nr;
+
+	if (flags & DAOS_ZERO_COPY) {
+		dtsr->dtsr_zero_copy = 1;
+		dtsr->dtsr_dkey = dkey;
+		dtsr->dtsr_akeys = akeys;
+	} else {
+		dtsr->dtsr_dkey = &dtsr->dtsr_dkey_inline;
+		rc = daos_iov_copy(dtsr->dtsr_dkey, &dkey[0]);
+		if (rc != 0)
+			D_GOTO(out, rc);
+
+		if (nr == 0)
+			goto link;
+
+		D_ALLOC_ARRAY(dtsr->dtsr_akeys, nr);
+		if (dtsr->dtsr_akeys == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		for (i = 0; i < nr; i++) {
+			rc = daos_iov_copy(&dtsr->dtsr_akeys[i], &akeys[i]);
+			if (rc != 0)
+				D_GOTO(out, rc);
+		}
+	}
+
+link:
+	d_list_add_tail(&dtsr->dtsr_link, &tx->tx_sub_reqs);
+	tx->tx_sub_count++;
+
+out:
+	dc_tx_decref(tx);
+	if (rc != 0 && dtsr != NULL) {
+		if (dtsr->dtsr_obj != NULL)
+			obj_decref(dtsr->dtsr_obj);
+
+		if (!(flags & DAOS_ZERO_COPY)) {
+			if (dtsr->dtsr_akeys != NULL) {
+				for (i = 0; i < nr; i++)
+					daos_iov_free(&dtsr->dtsr_akeys[i]);
+
+				D_FREE(dtsr->dtsr_akeys);
+			}
+
+			daos_iov_free(dtsr->dtsr_dkey);
+		}
+
+		D_FREE(dtsr);
+	}
+
+	return rc;
+}
+
+int
+dc_tx_attach(void *args, enum obj_rpc_opc opc)
+{
+	switch (opc) {
+	case DAOS_OBJ_RPC_UPDATE: {
+		daos_obj_update_t	*up = args;
+
+		return dc_tx_update_attach(up->oh, up->th, up->flags, up->dkey,
+					   up->nr, up->iods, up->sgls);
+	}
+	case DAOS_OBJ_RPC_PUNCH: {
+		daos_obj_punch_t	*pu = args;
+
+		return dc_tx_punch_attach(pu->oh, pu->th, pu->flags);
+	}
+	case DAOS_OBJ_RPC_PUNCH_DKEYS: {
+		daos_obj_punch_t	*pu = args;
+
+		return dc_tx_punch_dkeys_attach(pu->oh, pu->th, pu->flags,
+						pu->dkey);
+	}
+	case DAOS_OBJ_RPC_PUNCH_AKEYS: {
+		daos_obj_punch_t	*pu = args;
+
+		return dc_tx_punch_akeys_attach(pu->oh, pu->th, pu->flags,
+						pu->dkey, pu->akey_nr,
+						pu->akeys);
+	}
+	default:
```

Comment:
(style)  Possible switch case/default not preceeded by break or fallthrough comment